### DISCUSSION TITLE
sve  and avx512 masked  intrinsics and sve semi-scalar  intrinsics

### DISCUSSION THREAD

**Discussion by jtlap** [jfalcou/eve#1846] [Ideas]
sve  and avx512 provides many masked intrinsics with 2 or 3 flavours up  to now we only use the first one

 - merge with first input:  m
 - merge with 0: z
 - merge with unknown: u (sve)
 
perhaps we have to also support some way the last 2 items with a syntax that could be
pipo[<mode>(mask)](...)  or pipo[mode] [mask)](...) where mode could be m z or u with pipo[mask](...)   meaning m.
Perhaps m z and u could be  upper_case (M,Z,U) to avoid most ambiguities 

Also sve provides many mixed vector scalar intrinsics. The elementwise_callable way we define many functors does not allow a call to these intrinsics as pipo(vec,scalar) is converted in pipo(vec,brodcast(scalar) before the call the intrinsic.
Is there something to do ans is the gain sufficient to need an effort ?

**Comment by jfalcou** [comment]
>     * merge with 0: z

THis probabyl just need us to special case `func[if_(mask).else_(eve::zero)]` to be treated specially.
No this for additional markers.

>     * merge with unknown: u (sve)
Having unspecified values is probably not what we want. Testing will be hell.

> Also sve provides many mixed vector scalar intrinsics. The elementwise_callable way we define many functors does not allow a call to these intrinsics as pipo(vec,scalar) is converted in pipo(vec,brodcast(scalar) before the call the intrinsic. Is there something to do ans is the gain sufficient to need an effort ?

THat is what `strict_elementwise_callable`is for : let you define fine grained overload. No need for more.

**Comment by jtlap** [comment]
ok just that func[if_(mask).else_(eve::zero)]  is a little heavy

**Comment by DenisYaroshevskiy** [comment]
I also think that we need to make the compiler optimize those a bit
